cdci_data_analysis/analysis/email_helper.py

```diff
@@ -368,65 +377,82 @@ def send_email(smtp_server,
-            server.quit()
-
-    logger.info("email successfully sent")
+    n_tries_left = num_email_sending_max_tries
+
+    if not isinstance(receiver_email_addresses, list):
+        receiver_email_addresses = [receiver_email_addresses]
+    if cc_receivers_email_addresses is None:
+        cc_receivers_email_addresses = []
+    if bcc_receivers_email_addresses is None:
+        bcc_receivers_email_addresses = []
+    # include bcc receivers, which will be hidden in the message header
+    receivers_email_addresses = receiver_email_addresses + cc_receivers_email_addresses + bcc_receivers_email_addresses
+    # creation of the message
+    message = MIMEMultipart("alternative")
+    message["Subject"] = email_subject
+    message["From"] = sender_email_address
+    message["To"] = ", ".join(receiver_email_addresses)
+    message["CC"] = ", ".join(cc_receivers_email_addresses)
+    message['Reply-To'] = reply_to_email_address
+
+    if attachment is not None:
+        # create the attachment
+        message.attach(attachment)
+
+    part1 = MIMEText(email_text, "plain")
+    part2 = MIMEText(email_body_html, "html")
+    message.attach(part1)
+    message.attach(part2)
 
-    return message
+    while True:
+        try:
+            # Create a secure SSL context
+            context = ssl.create_default_context()
+            #
+            # Try to log in to server and send email
+            server = smtplib.SMTP(smtp_server, smtp_port)
+            # just for testing purposes, not ssl is established
+            if smtp_server != "localhost":
+                try:
+                    server.starttls(context=context)
+                except Exception as e:
+                    logger.warning(f'unable to start TLS: {e}')
+            if smtp_server_password is not None and smtp_server_password != '':
+                server.login(sender_email_address, smtp_server_password)
+            server.sendmail(sender_email_address, receivers_email_addresses, message.as_string())
+            logger.info("email successfully sent")
+
+            return message
+        except Exception as e:
+            n_tries_left -= 1
+
+            if n_tries_left > 0:
+                logger.warning(f"there seems to be some problem in sending the email with title {email_subject}, "
+                               f"another attempt will be made")
+
+                logger.debug(f"{e} exception while attempting to send the email with title {email_subject}\n"
+                             f"{n_tries_left} tries left, sleeping {email_sending_retry_sleep_s} seconds until retry\n")
+                time.sleep(email_sending_retry_sleep_s)
+            else:
+                logger.warning(f"an issue occurred when sending the email with title {email_subject}, "
+                               f"we are investigating and try to solve the issue as soon as possible")
+
+                store_not_sent_email(email_body_html, scratch_dir, sending_time=sending_time)
```

Comment:
I would capture sentry here to be sure. Even if it also caputres elsewhere.

Reply:
but a sentry is already captured at upper levels, we would end up with an additional sentry log that would not add much to what we already receive

Reply:
yes, that's why I am saying that I would capture sentry here to be sure. Even if it also captures elsewhere.
Clearly, there is a problem with what sentry captures and where. Maybe there is a mistake somewhere.
So, to be sure, we can also capture here. In the worst case we get two messages with slightly different context.